### ISSUE TITLE
Getting Error when running validating xmlschema in threads

### ISSUE THREAD

**Issue by arifshaon** [sissaschool/xmlschema#147]
Hello,

I am not sure what is causing this but  I am getting the following error when I am trying to validate multiple XML instances against the same XSD using threads.

global simpleType with name=u'{http://www.w3.org/2001/XMLSchema}formChoice' is already defined.

Please advise.

Thanks
Arif

**Comment by brunato** [comment]
Hi,
could you provide a sample of the code that you use for schema validation with threads?
Thanks

**Comment by arifshaon** [comment]
Sure.

**The code the validates XML runs in a function like the one below:**
XSD_PATH = "http://www.loc.gov/standards/vracore/vra-strict.xsd"

define validate_xml(xml_path):
  try:
     
     xsd = xmlschema.XMLSchema(XSD_PATH)
     xsd.validate(xml_path)
  except xmlschema.validators.exceptions.XMLSchemaValidationError as ve:
     print "handle validation error"
  except Exception as e:
     print "handle parsing error" 
 
**The code that runs this in a queue**

q = LifoQueue(maxsize=0)

for xml_path in xml_files:
  q.put((validate_xml, [xml_path]))

Please do let me know if you need additional information.

Cheers

**Comment by brunato** [comment]
Hi,
the problem is in schema building because the threads are activated to build the same meta-schema objects, so these build phases fail. For making the schema build compatible with threads there are several changes and tests to do.

A workarount with the current version is to create the schema once and then validate the XML files in threads, that is also less resource/time consuming. For example:

```python
XSD_PATH = "http://www.loc.gov/standards/vracore/vra-strict.xsd"

xsd = xmlschema.XMLSchema(XSD_PATH)

define validate_xml(xml_path):
  try:
     xsd.validate(xml_path)
  except xmlschema.validators.exceptions.XMLSchemaValidationError as ve:
     print "handle validation error"
  except Exception as e:
     print "handle parsing error" 
```

**Comment by brunato** [comment]
Fix available with v1.0.17+.
I've added a minimal support (locks on building maps and on XPath operations) for threads, so you now you can share global maps between more schemas on different threads without building errors.
Bye
